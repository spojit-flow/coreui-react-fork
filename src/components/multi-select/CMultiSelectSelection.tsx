import React, { forwardRef, HTMLAttributes } from 'react'

import classNames from 'classnames'
import PropTypes from 'prop-types'

import type { SelectedOption } from './types'

export interface CMultiSelectSelectionProps extends HTMLAttributes<HTMLSpanElement> {
  multiple?: boolean
  onRemove?: (option: SelectedOption) => void
  placeholder?: string
  search?: boolean | 'external'
  selected?: SelectedOption[]
  selectionType?: 'counter' | 'tags' | 'text'
  selectionTypeCounterText?: string
}

export const CMultiSelectSelection = forwardRef<HTMLSpanElement, CMultiSelectSelectionProps>(
  (
    {
      children,
      multiple,
      placeholder,
      onRemove,
      search,
      selected = [],
      selectionType,
      selectionTypeCounterText,
    },
    ref,
  ) => {
    return (
      <span
        className={classNames('form-multi-select-selection', {
          'form-multi-select-selection-tags': multiple && selectionType === 'tags',
        })}
        ref={ref}
      >
        {multiple && selectionType === 'counter' && !search && selected.length === 0 && placeholder}
        {multiple &&
          selectionType === 'counter' &&
          !search &&
          selected.length > 0 &&
          `${selected.length} ${selectionTypeCounterText}`}
        {multiple &&
          selectionType === 'tags' &&
          selected.map((option: SelectedOption, index: number) => {
            if (selectionType === 'tags') {
              return (
                <span className="form-multi-select-tag" key={index}>
                  {option.text}
                  {!option.disabled && (
                    <button
                      className="form-multi-select-tag-delete"
                      aria-label="Close"
                      onClick={() => onRemove && onRemove(option)}
                    >
                      <span aria-hidden="true">Ã—</span>
                    </button>
                  )}
                </span>
              )
            }
            return
          })}
        {multiple &&
          selectionType === 'text' &&
          selected.map((option, index) => (
            <span key={index}>
              {option.text}
              {index === selected.length - 1 ? '' : ','}&nbsp;
            </span>
          ))}
        {!multiple && !search && selected.map((option) => option.text)[0]}
        {children}
      </span>
    )
  },
)

CMultiSelectSelection.propTypes = {
  multiple: PropTypes.bool,
  onRemove: PropTypes.func,
  placeholder: PropTypes.string,
  search: PropTypes.oneOfType([PropTypes.bool, PropTypes.oneOf<'external'>(['external'])]),
  selected: PropTypes.array,
  selectionType: PropTypes.oneOf(['counter', 'tags', 'text']),
  selectionTypeCounterText: PropTypes.string,
}

CMultiSelectSelection.displayName = 'CMultiSelectSelection'
