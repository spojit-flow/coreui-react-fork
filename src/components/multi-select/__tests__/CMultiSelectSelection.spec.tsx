import * as React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { CMultiSelectSelection } from '../CMultiSelectSelection'

const selected = [
  {
    text: 'text1',
    value: 'value1',
  },
  {
    text: 'text2',
    value: 'value2',
    disabled: true,
  },
]

test('loads and displays CMultiSelectSelection component', async () => {
  const { container } = render(<CMultiSelectSelection />)
  expect(container).toMatchSnapshot()
})

test('CMultiSelectSelection customize tags', async () => {
  const { container } = render(
    <CMultiSelectSelection
      multiple={true}
      search={true}
      selected={selected}
      selectionType="tags"
      selectionTypeCounterText="text"
    />,
  )
  expect(container).toMatchSnapshot()
  expect(container.firstChild).toHaveClass('form-multi-select-selection')
  if (container.firstChild === null) {
    expect(true).toBe(false)
  } else {
    expect(container.firstChild.firstChild).toHaveClass('form-multi-select-tag')
    expect(container.firstChild.lastChild).toHaveClass('form-multi-select-tag')
    if (container.firstChild.firstChild === null) {
      expect(true).toBe(false)
    } else {
      expect(container.firstChild.firstChild.lastChild).toHaveClass('form-multi-select-tag-delete')
      expect(container.firstChild.firstChild.lastChild).toHaveAttribute('aria-label', 'Close')
      if (container.firstChild.firstChild.lastChild === null) {
        expect(true).toBe(false)
      } else {
        expect(container.firstChild.firstChild.lastChild.firstChild).toHaveAttribute(
          'aria-hidden',
          'true',
        )
        expect(container.firstChild.firstChild.lastChild.firstChild).toHaveTextContent('Ã—')
      }
    }
  }
})

test('CMultiSelectSelection customize counter', async () => {
  const { container } = render(
    <CMultiSelectSelection
      multiple={true}
      search={false}
      selected={selected}
      selectionType="counter"
      selectionTypeCounterText="text"
    />,
  )
  expect(container).toMatchSnapshot()
  expect(container.firstChild).toHaveClass('form-multi-select-selection')
  expect(container.firstChild).toHaveTextContent('2 text')
})

test('CMultiSelectSelection customize text', async () => {
  const { container } = render(
    <CMultiSelectSelection
      multiple={true}
      search={true}
      selected={selected}
      selectionType="text"
      selectionTypeCounterText="text"
    />,
  )
  expect(container).toMatchSnapshot()
  expect(container.firstChild).toHaveClass('form-multi-select-selection')
  expect(container.firstChild).toHaveTextContent('text1, text2')
})

test('CMultiSelectSelection test remove event', async () => {
  const onRemove = jest.fn()
  render(
    <CMultiSelectSelection
      multiple={true}
      selected={selected}
      selectionType="tags"
      onRemove={onRemove}
    />,
  )
  expect(onRemove).toHaveBeenCalledTimes(0)
  const removeButton = document.querySelector('.form-multi-select-tag-delete')
  if (removeButton !== null) {
    fireEvent.click(removeButton)
  }
  expect(onRemove).toHaveBeenCalledTimes(1)
})
