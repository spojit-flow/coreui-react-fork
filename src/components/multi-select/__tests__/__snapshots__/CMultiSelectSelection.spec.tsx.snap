// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CMultiSelectSelection customize counter 1`] = `
<div>
  <span
    class="form-multi-select-selection"
  >
    2 text
  </span>
</div>
`;

exports[`CMultiSelectSelection customize tags 1`] = `
<div>
  <span
    class="form-multi-select-selection form-multi-select-selection-tags"
  >
    <span
      class="form-multi-select-tag"
    >
      text1
      <button
        aria-label="Close"
        class="form-multi-select-tag-delete"
      >
        <span
          aria-hidden="true"
        >
          Ã—
        </span>
      </button>
    </span>
    <span
      class="form-multi-select-tag"
    >
      text2
    </span>
  </span>
</div>
`;

exports[`CMultiSelectSelection customize text 1`] = `
<div>
  <span
    class="form-multi-select-selection"
  >
    text1, text2
  </span>
</div>
`;

exports[`loads and displays CMultiSelectSelection component 1`] = `
<div>
  <span
    class="form-multi-select-selection"
  />
</div>
`;
